src/wp-admin/includes/class-wp-automatic-updater.php

```diff
@@ -1551,4 +1678,77 @@ protected function send_debug_email() {
 
 		wp_mail( $email['to'], wp_specialchars_decode( $email['subject'] ), $email['body'], $email['headers'] );
 	}
+
+	/**
+	 * Performs a loopback request to check for potential fatal errors.
+	 *
+	 * Fatal errors cannot be detected unless maintenance mode is enabled.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @global int $upgrading The Unix timestamp marking when upgrading WordPress began.
+	 *
+	 * @return bool Whether a fatal error was detected.
+	 */
+	protected function has_fatal_error() {
+		global $upgrading;
+
+		$maintenance_file = ABSPATH . '.maintenance';
+		if ( ! file_exists( $maintenance_file ) ) {
+			return false;
+		}
+
+		require $maintenance_file;
+		if ( ! is_int( $upgrading ) ) {
+			return false;
+		}
+
+		$scrape_key   = md5( $upgrading );
+		$scrape_nonce = (string) $upgrading;
+		$transient    = 'scrape_key_' . $scrape_key;
+		set_transient( $transient, $scrape_nonce, 30 );
+
+		$cookies       = wp_unslash( $_COOKIE );
+		$scrape_params = array(
+			'wp_scrape_key'   => $scrape_key,
+			'wp_scrape_nonce' => $scrape_nonce,
+		);
+		$headers       = array(
+			'Cache-Control' => 'no-cache',
+		);
+
+		/** This filter is documented in wp-includes/class-wp-http-streams.php */
+		$sslverify = apply_filters( 'https_local_ssl_verify', false );
+
+		// Include Basic auth in the loopback request.
+		if ( isset( $_SERVER['PHP_AUTH_USER'] ) && isset( $_SERVER['PHP_AUTH_PW'] ) ) {
+			$headers['Authorization'] = 'Basic ' . base64_encode( wp_unslash( $_SERVER['PHP_AUTH_USER'] ) . ':' . wp_unslash( $_SERVER['PHP_AUTH_PW'] ) );
+		}
+
+		// Time to wait for loopback request to finish.
+		$timeout = 50; // 50 seconds.
+
+		error_log( '    Scraping home page...' );
+
+		$needle_start = "###### wp_scraping_result_start:$scrape_key ######";
+		$needle_end   = "###### wp_scraping_result_end:$scrape_key ######";
+		$url          = add_query_arg( $scrape_params, home_url( '/' ) );
+		$response     = wp_remote_get( $url, compact( 'cookies', 'headers', 'timeout', 'sslverify' ) );
+
```

Comment:
```suggestion

		if ( is_wp_error( $response ) ) {
			$response = (array) $response;
			$response['body'][] = $needle_start;
			$response['body'][] = json_encode(
				array(
					'type'    => '1',
					'message' => 'Loopback is having a problem and any auto-updates will be rollback for safety.',
				)
			);
			$response['body'][] = $needle_end;
			$response['body']   = implode( "\n", $response['body'] );
		}

```

Reply:
This tests out fine on my end.

Reply:
It seems like this could be a shorthand of the above:

```php
if ( is_wp_error( $response ) ) {
    error_log( 'Loopback is having a problem and the update will be rolled back for safety.' );
    return true;
}
```

We can uncouple the message from rollback, and ensure the error from the response is included in the output, such as:

```php
error_log( 'Loopback request failed: ' . $response->get_error_message() );
```

Reply:
Yeah that's right, it's a replacement for the error message.

Reply:
Let's go with the shortcut and finish off the merge proposal so we can watch all the testing being done in the following 2 weeks and then commit if no issues reported.

Reply:
Don’t we need to report that the update was rolled back?

Reply:
Yep, but only in the _usage_ of `has_fatal_error()` (when the rollback is performed) and in the update email.

Reply:
The shortcut won't have any messaging about the rollback. Only the email.

Reply:
That's right yeah. The method's used as a component of Rollback Auto-Update, but it's a useful piece of functionality on it's own, so we should try to remove mention of Rollback inside it so it has potential for use elsewhere.

Rather than mention rolling back in `::has_fatal_error()`, which is about trying to detect a fatal error, we instead already mention rolling back in `::update()` via the existing `WP_Error` messages, where we act based on the result of `::has_fatal_error()`.

Reply:
Suggestion updated to correspond to ongoing discussion.